numpy/distutils/command/build_clib.py

```diff
@@ -289,16 +289,22 @@ def build_a_library(self, build_info, lib_name, libraries):
             # problem, msvc uses its own convention :(
             c_sources += cxx_sources
             cxx_sources = []
-
-        objects = []
+            obj_ext = '.obj'
+        else:
+            obj_ext = '.o'
+        
+        # c_sources collects all the default sources, including pre-compiled
+        # object files
+        objects = [c_sources.pop(c_sources.index(c)) for c in c_sources[:]
+                   if c.endswith(obj_ext)]
```

Comment:
Yuk. sorry for this, if there is a better way please suggest it. The `c_sources` is from
```
c_sources, cxx_sources, f_sources, fmodule_sources \
            = filter_sources(sources)
```
above, and is a bit misnamed since it collects anything that is not `cxx_sources`, `f_source` or `fmodule_sources`. Also see line 264 where `dispatch_sources` does something similar

Reply:
Can this be done instead by passing in a `.lib` file with `libraries + library_dirs`, so distutils hacking is not needed?

Reply:
Also, can the `.lib` file be compiled from sources with MASM in the `setup.py` file, so that it doesn't need to be shipped with Numpy?

Reply:
good idea, thanks

Reply:
I think this will require more hacking than the current solution. `setup.py` is not really the place to be running compilation, that is the job of `config.add_library` and friends. Is shipping a 1k binary object file that is easily dissasembled a deal breaker?

Reply:
Is there an example around of using MASM from python in a setup.py ? Maybe I am not seeing something obvious.

Reply:
Ahh, got it. There is an `extra_objects` arg to `Extensions`

Reply:
I was hoping it would be simple to do with running via `subprocess.run` the commands that you would do manually, i.e., minimize interfacing with distutils at all. And passing the generated file to the Extension.

I'd guess shipping binary blobs is not a dealbreaker, given that hopefully this hack can be dropped at some point in future, after the CRT bug is fixed.

Reply:
Maybe [pycca](https://github.com//pycca/pycca) can be used for that assembler snippet, bit it would be another dependency.